File: examples/canonical-terminal/python-typer-textual/taskflow/core/stats.py
from __future__ import annotations

from taskflow.core.models import Job, JobStatus, Stats


def compute_stats(jobs: list[Job]) -> Stats:
    total = len(jobs)
    by_status = {status.value: 0 for status in JobStatus}
    durations: list[float] = []

    for job in jobs:
        by_status[job.status.value] += 1
        if job.duration_s is not None:
            durations.append(job.duration_s)

    avg_duration = round(sum(durations) / len(durations), 2) if durations else 0.0
    failure_rate = round((by_status[JobStatus.FAILED.value] / total), 2) if total else 0.0
    return Stats(
        total=total,
        by_status=by_status,
        avg_duration_s=avg_duration,
        failure_rate=failure_rate,
    )

File: examples/canonical-terminal/rust-argh-tui-realm/src/core/stats.rs
use std::collections::HashMap;

use crate::core::models::{Job, JobStatus, Stats};

pub fn compute_stats(jobs: &[Job]) -> Stats {
    let mut by_status: HashMap<String, usize> = HashMap::from([
        (JobStatus::Pending.as_str().to_string(), 0),
        (JobStatus::Running.as_str().to_string(), 0),
        (JobStatus::Done.as_str().to_string(), 0),
        (JobStatus::Failed.as_str().to_string(), 0),
    ]);
    let mut durations = Vec::new();
    for job in jobs {
        *by_status
            .entry(job.status.as_str().to_string())
            .or_default() += 1;
        if let Some(duration) = job.duration_s {
            durations.push(duration);
        }
    }
    let total = jobs.len();
    let avg_duration_s = if durations.is_empty() {
        0.0
    } else {
        let sum: f64 = durations.iter().sum();
        (sum / durations.len() as f64 * 100.0).round() / 100.0
    };
    let failed = *by_status.get("failed").unwrap_or(&0);
    let failure_rate = if total == 0 {
        0.0
    } else {
        ((failed as f64 / total as f64) * 100.0).round() / 100.0
    };
    Stats {
        total,
        by_status,
        avg_duration_s,
        failure_rate,
    }
}
